File: src/test/java/cube8540/oauth/authentication/credentials/resource/application/DefaultSecuredResourceManagementServiceTest.java
package cube8540.oauth.authentication.credentials.resource.application;

import cube8540.oauth.authentication.credentials.resource.domain.ResourceMethod;
import cube8540.oauth.authentication.credentials.resource.domain.ResourceNotFoundException;
import cube8540.oauth.authentication.credentials.resource.domain.ResourceRegisterException;
import cube8540.oauth.authentication.credentials.resource.domain.SecuredResource;
import cube8540.oauth.authentication.credentials.resource.domain.SecuredResourceRepository;
import cube8540.oauth.authentication.credentials.resource.domain.SecuredResourceValidatorFactory;
import cube8540.oauth.authentication.error.message.ErrorCodes;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.InOrder;

import static cube8540.oauth.authentication.credentials.resource.application.SecuredResourceApplicationTestHelper.ADD_AUTHORITIES;
import static cube8540.oauth.authentication.credentials.resource.application.SecuredResourceApplicationTestHelper.ADD_REQUEST_AUTHORITIES;
import static cube8540.oauth.authentication.credentials.resource.application.SecuredResourceApplicationTestHelper.AUTHORITIES;
import static cube8540.oauth.authentication.credentials.resource.application.SecuredResourceApplicationTestHelper.MODIFY_RESOURCE_URI;
import static cube8540.oauth.authentication.credentials.resource.application.SecuredResourceApplicationTestHelper.RAW_MODIFY_RESOURCE_URI;
import static cube8540.oauth.authentication.credentials.resource.application.SecuredResourceApplicationTestHelper.RAW_RESOURCE_ID;
import static cube8540.oauth.authentication.credentials.resource.application.SecuredResourceApplicationTestHelper.RAW_RESOURCE_URI;
import static cube8540.oauth.authentication.credentials.resource.application.SecuredResourceApplicationTestHelper.REMOVE_AUTHORITIES;
import static cube8540.oauth.authentication.credentials.resource.application.SecuredResourceApplicationTestHelper.REMOVE_REQUEST_AUTHORITIES;
import static cube8540.oauth.authentication.credentials.resource.application.SecuredResourceApplicationTestHelper.REQUEST_AUTHORITIES;
import static cube8540.oauth.authentication.credentials.resource.application.SecuredResourceApplicationTestHelper.RESOURCE_ID;
import static cube8540.oauth.authentication.credentials.resource.application.SecuredResourceApplicationTestHelper.RESOURCE_URI;
import static cube8540.oauth.authentication.credentials.resource.application.SecuredResourceApplicationTestHelper.makeDefaultSecuredResource;
import static cube8540.oauth.authentication.credentials.resource.application.SecuredResourceApplicationTestHelper.makeEmptyResourceRepository;
import static cube8540.oauth.authentication.credentials.resource.application.SecuredResourceApplicationTestHelper.makeErrorValidatorFactory;
import static cube8540.oauth.authentication.credentials.resource.application.SecuredResourceApplicationTestHelper.makeResourceRepository;
import static cube8540.oauth.authentication.credentials.resource.application.SecuredResourceApplicationTestHelper.makeValidatorFactory;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.inOrder;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

@DisplayName("기본 자원 관리 서비스 테스트")
class DefaultSecuredResourceManagementServiceTest {

    @Test
    @DisplayName("이미 등록된 리소스 영속화")
    void persistResourceToAlreadyRegisteredInRepository() {
        SecuredResourceRegisterRequest request = new SecuredResourceRegisterRequest(RAW_RESOURCE_ID, RAW_RESOURCE_URI, "POST", null);
        SecuredResource resource = makeDefaultSecuredResource();
        SecuredResourceRepository repository = makeResourceRepository(RESOURCE_ID, resource);
        DefaultSecuredResourceManagementService service = new DefaultSecuredResourceManagementService(repository);

        ResourceRegisterException exception = assertThrows(ResourceRegisterException.class, () -> service.registerNewResource(request));
        assertEquals(ErrorCodes.EXISTS_IDENTIFIER, exception.getCode());
    }

    @Test
    @DisplayName("유효 하지 않은 리소스 데이터 등록")
    void registerInvalidResource() {
        SecuredResourceRegisterRequest request = new SecuredResourceRegisterRequest(RAW_RESOURCE_ID, RAW_RESOURCE_URI, "POST", REQUEST_AUTHORITIES);
        SecuredResourceRepository repository = makeEmptyResourceRepository();
        SecuredResourceValidatorFactory factory = makeErrorValidatorFactory(new TestSecuredResourceException());
        DefaultSecuredResourceManagementService service = new DefaultSecuredResourceManagementService(repository);

        service.setValidatorFactory(factory);

        assertThrows(TestSecuredResourceException.class, () -> service.registerNewResource(request));
    }

    @Test
    @DisplayName("새 리소스 등록")
    void registerNewResource() {
        ArgumentCaptor<SecuredResource> resourceCaptor = ArgumentCaptor.forClass(SecuredResource.class);
        SecuredResourceRegisterRequest request = new SecuredResourceRegisterRequest(RAW_RESOURCE_ID, RAW_RESOURCE_URI, "POST", REQUEST_AUTHORITIES);
        SecuredResourceRepository repository = makeEmptyResourceRepository();
        SecuredResourceValidatorFactory factory = makeValidatorFactory();
        DefaultSecuredResourceManagementService service = new DefaultSecuredResourceManagementService(repository);

        service.setValidatorFactory(factory);
        service.registerNewResource(request);

        verify(repository, times(1)).save(resourceCaptor.capture());
        assertEquals(RESOURCE_ID, resourceCaptor.getValue().getResourceId());
        assertEquals(RESOURCE_URI, resourceCaptor.getValue().getResource());
        assertEquals(ResourceMethod.POST, resourceCaptor.getValue().getMethod());
        assertEquals(AUTHORITIES, resourceCaptor.getValue().getAuthorities());
    }

    @Test
    @DisplayName("저장소에 등록 되지 않은 리소스 수정")
    void modifyNotRegisteredResourceInRepository() {
        SecuredResourceModifyRequest request = new SecuredResourceModifyRequest(RAW_MODIFY_RESOURCE_URI, "PUT", null, null);
        SecuredResourceRepository repository = makeEmptyResourceRepository();
        DefaultSecuredResourceManagementService service = new DefaultSecuredResourceManagementService(repository);

        ResourceNotFoundException exception = assertThrows(ResourceNotFoundException.class, () -> service.modifyResource(RAW_RESOURCE_ID, request));
        assertEquals(ErrorCodes.NOT_FOUND, exception.getCode());
    }

    @Test
    @DisplayName("유효 하지 않은 데이터로 리소스 수정")
    void modifyResourceWithInvalidData() {
        SecuredResourceModifyRequest request = new SecuredResourceModifyRequest(RAW_MODIFY_RESOURCE_URI, "PUT", ADD_REQUEST_AUTHORITIES, REMOVE_REQUEST_AUTHORITIES);
        SecuredResource resource = makeDefaultSecuredResource();
        SecuredResourceRepository repository = makeResourceRepository(RESOURCE_ID, resource);
        SecuredResourceValidatorFactory factory = makeErrorValidatorFactory(new TestSecuredResourceException());
        DefaultSecuredResourceManagementService service = new DefaultSecuredResourceManagementService(repository);

        service.setValidatorFactory(factory);
        service.modifyResource(RAW_RESOURCE_ID, request);

        InOrder inOrder = inOrder(resource, repository);
        inOrder.verify(resource, times(1)).validation(factory);
        inOrder.verify(repository, times(1)).save(resource);
    }

    @Test
    @DisplayName("리소스 수정")
    void modifyResource() {
        SecuredResourceModifyRequest request = new SecuredResourceModifyRequest(RAW_MODIFY_RESOURCE_URI, "PUT", ADD_REQUEST_AUTHORITIES, REMOVE_REQUEST_AUTHORITIES);
        SecuredResource resource = makeDefaultSecuredResource();
        SecuredResourceRepository repository = makeResourceRepository(RESOURCE_ID, resource);
        SecuredResourceValidatorFactory factory = makeValidatorFactory();
        DefaultSecuredResourceManagementService service = new DefaultSecuredResourceManagementService(repository);

        service.setValidatorFactory(factory);

        service.modifyResource(RAW_RESOURCE_ID, request);
        InOrder inOrder = inOrder(resource, repository);
        inOrder.verify(resource, times(1)).changeResourceInfo(MODIFY_RESOURCE_URI, ResourceMethod.PUT);
        REMOVE_AUTHORITIES.forEach(auth -> inOrder.verify(resource, times(1)).removeAuthority(auth.getAuthority()));
        ADD_AUTHORITIES.forEach(auth -> inOrder.verify(resource, times(1)).addAuthority(auth.getAuthority()));
        inOrder.verify(repository, times(1)).save(resource);
    }

    @Test
    @DisplayName("저장소에 등록 되지 않은 리소스 삭제")
    void removeNotRegisteredResourceInRepository() {
        SecuredResourceRepository repository = makeEmptyResourceRepository();
        DefaultSecuredResourceManagementService service = new DefaultSecuredResourceManagementService(repository);

        ResourceNotFoundException exception = assertThrows(ResourceNotFoundException.class, () -> service.removeResource(RAW_RESOURCE_ID));
        assertEquals(ErrorCodes.NOT_FOUND, exception.getCode());
    }

    @Test
    @DisplayName("리소스 삭제")
    void removeResource() {
        SecuredResource resource = makeDefaultSecuredResource();
        SecuredResourceRepository repository = makeResourceRepository(RESOURCE_ID, resource);
        DefaultSecuredResourceManagementService service = new DefaultSecuredResourceManagementService(repository);

        service.removeResource(RAW_RESOURCE_ID);
        verify(repository, times(1)).delete(resource);
    }

    private static class TestSecuredResourceException extends RuntimeException {}
}

File: src/test/kotlin/cube8540/oauth/authentication/resource/application/DefaultSecuredResourceManagementServiceTest.kt
package cube8540.oauth.authentication.resource.application

import cube8540.oauth.authentication.UnitTestValidationException
import cube8540.oauth.authentication.resource.domain.*
import io.mockk.*
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.catchThrowable
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import java.net.URI
import java.util.*
import kotlin.random.Random

class DefaultSecuredResourceManagementServiceTest {

    private val repository: SecuredResourceRepository = mockk(relaxed = true)
    private val validatorFactory: SecuredResourceValidatorFactory = mockk()

    private val service: DefaultSecuredResourceManagementService = DefaultSecuredResourceManagementService(repository)

    init {
        every { repository.save(any()) } returnsArgument 0

        service.validatorFactory = validatorFactory
    }

    @Nested
    inner class CountingTest {

        @Test
        fun `resource counting`() {
            val count = Random.nextLong(0, 100)

            every { repository.countByResourceId(SecuredResourceId("resourceId")) } returns count

            val results = service.count("resourceId")
            assertThat(results).isEqualTo(count)
        }
    }

    @Nested
    inner class RegisterNewResourceTest {

        @Test
        fun `resource is already registered in repository`() {
            val accessibleAuthorities = listOf(AccessibleAuthorityValue("access.test.1"), AccessibleAuthorityValue("access.test.2"))
            val request = SecuredResourceRegisterRequest("resourceId", "http://localhost:8080", "POST", accessibleAuthorities)

            every { repository.countByResourceId(SecuredResourceId("resourceId")) } returns 1

            val thrown = catchThrowable { service.registerNewResource(request) }
            assertThat(thrown).isInstanceOf(ResourceRegisterException::class.java)
        }

        @Test
        fun `request resource data is invalid`() {
            val accessibleAuthorities = listOf(AccessibleAuthorityValue("access.test.1"), AccessibleAuthorityValue("access.test.2"))
            val request = SecuredResourceRegisterRequest("resourceId", "http://localhost:8080", "POST", accessibleAuthorities)

            every { repository.countByResourceId(SecuredResourceId("resourceId")) } returns 0
            every { validatorFactory.createValidator(any()) } returns mockk {
                every { result } returns mockk {
                    every { hasErrorThrows(any()) } throws UnitTestValidationException()
                }
            }

            val thrown = catchThrowable { service.registerNewResource(request) }
            assertThat(thrown).isInstanceOf(UnitTestValidationException::class.java)
        }

        @Test
        fun `register successful`() {
            val resourceCaptor = slot<SecuredResource>()
            val accessibleAuthorities = listOf(AccessibleAuthorityValue("access.test.1"), AccessibleAuthorityValue("access.test.2"))
            val request = SecuredResourceRegisterRequest("resourceId", "http://localhost:8080", "POST", accessibleAuthorities)

            every { repository.countByResourceId(SecuredResourceId("resourceId")) } returns 0
            every { repository.save(capture(resourceCaptor)) } returnsArgument 0
            every { validatorFactory.createValidator(any()) } returns mockk {
                every { result } returns mockk {
                    every { hasErrorThrows(any()) } just Runs
                }
            }

            service.registerNewResource(request)
            assertThat(resourceCaptor.isCaptured).isTrue
            assertThat(resourceCaptor.captured.resourceId).isEqualTo(SecuredResourceId("resourceId"))
            assertThat(resourceCaptor.captured.resource).isEqualTo(URI.create("http://localhost:8080"))
            assertThat(resourceCaptor.captured.method).isEqualTo(ResourceMethod.POST)
            assertThat(resourceCaptor.captured.authorities)
                .isEqualTo(setOf(AccessibleAuthority("access.test.1"), AccessibleAuthority("access.test.2")))
        }
    }

    @Nested
    inner class ModifyResourceTest {

        @Test
        fun `request resource is not registered in repository`() {
            val newAuthorities = listOf(AccessibleAuthorityValue("new.access.1"), AccessibleAuthorityValue("new.access.2"))
            val removeAuthorities = listOf(AccessibleAuthorityValue("rem.access.1"), AccessibleAuthorityValue("rem.access.2"))
            val request = SecuredResourceModifyRequest("http://modify:8080", "PUT", newAuthorities, removeAuthorities)

            every { repository.findById(SecuredResourceId("resourceId")) } returns Optional.empty()

            val thrown = catchThrowable { service.modifyResource("resourceId", request) }
            assertThat(thrown).isInstanceOf(ResourceNotFoundException::class.java)
        }

        @Test
        fun `request resource data is invalid`() {
            val storedResource: SecuredResource = mockk(relaxed = true)
            val newAuthorities = listOf(AccessibleAuthorityValue("new.access.1"), AccessibleAuthorityValue("new.access.2"))
            val removeAuthorities = listOf(AccessibleAuthorityValue("rem.access.1"), AccessibleAuthorityValue("rem.access.2"))
            val request = SecuredResourceModifyRequest("http://modify:8080", "PUT", newAuthorities, removeAuthorities)

            every { repository.findById(SecuredResourceId("resourceId")) } returns Optional.of(storedResource)
            every { storedResource.validation(validatorFactory) } throws UnitTestValidationException()

            val thrown = catchThrowable { service.modifyResource("resourceId", request) }
            assertThat(thrown).isInstanceOf(UnitTestValidationException::class.java)
            verifyOrder {
                storedResource.changeResourceInfo(URI.create("http://modify:8080"), ResourceMethod.PUT)
                storedResource.removeAuthority("rem.access.1")
                storedResource.removeAuthority("rem.access.2")
                storedResource.addAuthority("new.access.1")
                storedResource.addAuthority("new.access.2")
                storedResource.validation(validatorFactory)
            }
        }

        @Test
        fun `modify successful`() {
            val savedResourceCaptor = slot<SecuredResource>()
            val storedResource: SecuredResource = mockk(relaxed = true)
            val newAuthorities = listOf(AccessibleAuthorityValue("new.access.1"), AccessibleAuthorityValue("new.access.2"))
            val removeAuthorities = listOf(AccessibleAuthorityValue("rem.access.1"), AccessibleAuthorityValue("rem.access.2"))
            val request = SecuredResourceModifyRequest("http://modify:8080", "PUT", newAuthorities, removeAuthorities)

            every { repository.findById(SecuredResourceId("resourceId")) } returns Optional.of(storedResource)
            every { storedResource.validation(validatorFactory) } just Runs

            service.modifyResource("resourceId", request)
            verifyOrder {
                storedResource.changeResourceInfo(URI.create("http://modify:8080"), ResourceMethod.PUT)
                storedResource.removeAuthority("rem.access.1")
                storedResource.removeAuthority("rem.access.2")
                storedResource.addAuthority("new.access.1")
                storedResource.addAuthority("new.access.2")
                storedResource.validation(validatorFactory)
                repository.save(capture(savedResourceCaptor))
            }
            assertThat(savedResourceCaptor.captured).isEqualTo(storedResource)
        }
    }

    @Nested
    inner class RemoveResourceTest {

        @Test
        fun `request resource is not registered in repository`() {
            every { repository.findById(SecuredResourceId("resourceId")) } returns Optional.empty()

            val thrown = catchThrowable { service.removeResource("resourceId") }
            assertThat(thrown).isInstanceOf(ResourceNotFoundException::class.java)
        }

        @Test
        fun `request resource is registered in repository`() {
            val resource: SecuredResource = mockk(relaxed = true)

            every { repository.findById(SecuredResourceId("resourceId")) } returns Optional.of(resource)

            service.removeResource("resourceId")
            verify { repository.delete(resource) }
        }
    }
}